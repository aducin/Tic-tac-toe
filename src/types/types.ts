interface BoxBase {
    icon: string;
    iconWinning: string;
    value: string;
};

export interface Box extends BoxBase {
    assigned: boolean;
    winning: boolean;
};

export type BoxesProps = {
    boxes: Box[];
    finished: boolean;
    onBoxClicked: (id: number) => void;
    started: boolean;
};

export type BoxItemProps = {
    data: {
        id: number;
        item: Box,
    };
    onBoxClicked: (id: number) => void;
}

export type ButtonsProps = {
    onButtonClicked: (button: string) => void;
    started: boolean;
};

export type DetailsProps = {
    activePlayer: string;
    turnTimer: number;
    value: string;
}

export type GameComponentState = {
    gameTimer: number;
    message: string;
    success: boolean;
    turnTimer: number;
    warning: boolean;
};

export type GameHistory = {
    duration: number;
    finishedTime: string;
    moves: number;
    winner: string;
};

export interface GameType extends BoxBase {
    id: number;
};

export type HistoryAction = {
    duration: number;
    finishedTime: Date;
}

export type IsPossibleToWin = {
    first: number;
    second: number
};

export type MatchedRows = {
    first: number;
    second: number;
};

export type MessageProps = {
    message: string;
    onRemoveAlert?: () => void;
    success: boolean;
    warning: boolean;
};

type Player = {
    active: boolean;
    name: string;
}

export type PlayerAction = {
    property: string;
    value: string;
}

export type PlayersProps = {
    onInputChanged: (property: string, name: string) => void;
    player1: string;
    player2: string;
}

export type ReducerAction = {
    payload: {
        boxes?: Box[];
        player?: PlayerAction;
        scoresList?: HistoryAction;
    }
    type: string;
};

export interface ScoresProps {
    scores: GameHistory[] | [];
}

export interface StoreState extends ScoresProps {
    boxes: Box[];
    finished: boolean;
    moves: number;
    next: GameType;
    player1: Player;
    player2: Player;
    started: boolean;
};
