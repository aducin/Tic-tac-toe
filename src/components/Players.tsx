import React from 'react';
import { PlayersProps } from '../interfaces/interfaces';
import styles from '../assets/styles/players.module.scss';

const Players: React.FC<PlayersProps> = (props: PlayersProps) => {
    const inputClasses = `form-control ${styles.input}`;
    return(
        <React.Fragment>
            <div className={styles.innerContainer}>
                <label className={styles.label}>Player 1</label>
                <input
                    onChange={(event) => props.onInputChanged('player1', event.target.value)}
                    className={inputClasses}
                    type="text"
                    value={props.player1}
                />
            </div>
            <div className={styles.innerContainer}>
                <label className={styles.label}>Player 2</label>
                <input
                    onChange={(event) => props.onInputChanged('player2', event.target.value)}
                    className={inputClasses}
                    type="text"
                    value={props.player2}
                />
            </div>
        </React.Fragment>
    );
}

export default React.memo(Players);