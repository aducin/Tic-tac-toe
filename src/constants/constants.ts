import { GameType } from '../interfaces/interfaces';

export const DEFAULT_PLAYER1_NAME = 'Player 1';
export const DEFAULT_PLAYER2_NAME = 'Player 2';
export const EMPTY_HISTORY = 'Game history is empty at the moment';
export const FOOTER_TEXT = 'Tic-tac-toe footer';
export const GAME_NOT_STARTED = 'Click the button to start the game';
export const GAME_OVER = 'Game is over - no winner';
export const GAME_SUCCEDED = (winner: string, duration: number) =>
    `${winner} has won in ${duration.toFixed(2)} seconds!`;
export const HEADER_TEXT = 'Tic-tac-toe header';
export const MESSAGE_HEADER_SUCCESS = 'CONGRATULATIONS!';
export const MESSAGE_HEADER_WARNING = 'WARNING!';
export const MESSAGE_TIME = 3000;
export const MOBILE_MAX_WIDTH = '768px';
export const PLAYER_FIRST_TYPE = 'O';
export const PLAYER_SECOND_TYPE = 'X';
export const ROWS_TO_WIN = 3;
export const SCORES_LIST_HEADER = 'Score results';
export const TIME_IS_OVER = 'Time is over - next move';
export const TURN_TIME = 30;
export const WINNING_ROWS = [
    [0, 1, 2],
    [3, 4, 5],
    [6, 7, 8],
    [0, 3, 6],
    [1, 4, 7],
    [2, 5, 8],
    [0, 4, 8],
    [2, 4, 6]
];
export const GAMETYPES: GameType[] = [
    { 
        id: 1,
        icon: 'ImageO',
        iconWinning: 'ImageOWinning',
        value: PLAYER_FIRST_TYPE
    },
    {
        id: 2,
        icon: 'ImageX',
        iconWinning: 'ImageXWinning',
        value: PLAYER_SECOND_TYPE
    }
];
