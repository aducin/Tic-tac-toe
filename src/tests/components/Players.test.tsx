import React from "react";
import TestRenderer from 'react-test-renderer';
import { shallow } from "enzyme";
import { ShallowWrapper } from 'enzyme';
import Players from '../../components/Players';
import { PlayersProps } from '../../types/types';
import { DEFAULT_PLAYER1_NAME , DEFAULT_PLAYER2_NAME } from '../../constants/constants';

const data: PlayersProps = {
    onInputChanged: () => {},
    player1: DEFAULT_PLAYER1_NAME,
    player2: DEFAULT_PLAYER2_NAME
}

const component = (
    <Players
        onInputChanged={data.onInputChanged}
        player1={data.player1}
        player2={data.player2}
    />
);

const testData = TestRenderer.create(component);

describe('Players component snapshot', () => {
    it('Players component renders correctly', async () => {
        expect(testData.toJSON()).toMatchSnapshot();
    });
});

describe('Players component tests', () => {
    let wrapper: ShallowWrapper;
    beforeEach(() => {
        wrapper = shallow(component);
    });

    it('there should be two input tags', () => {
        expect(wrapper.find('input')).toHaveLength(2);
    });

    it('there should be correct values in both inputs', () => {
        const el = wrapper.find('input');
        expect(el.first().prop('value')).toEqual(DEFAULT_PLAYER1_NAME);
        expect(el.last().prop('value')).toEqual(DEFAULT_PLAYER2_NAME);
    });
});

