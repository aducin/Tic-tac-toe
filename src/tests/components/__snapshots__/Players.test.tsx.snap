// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Players component snapshot Players component renders correctly 1`] = `
Array [
  <div
    className="innerContainer"
  >
    <label
      className="label"
    >
      Player 1
    </label>
    <input
      className="form-control input"
      onChange={[Function]}
      type="text"
      value="Player 1"
    />
  </div>,
  <div
    className="innerContainer"
  >
    <label
      className="label"
    >
      Player 2
    </label>
    <input
      className="form-control input"
      onChange={[Function]}
      type="text"
      value="Player 2"
    />
  </div>,
]
`;
